LogicLayer/Notifications/MailSender.cs

```diff
@@ -0,0 +1,115 @@
+                GetHtmlPath(MailOptions.EmailConfirmationHtml),
+                link,
+                "Email Confirmation");
+            return SendEmailAsync(email, notification);
+        }
+
+        public Task SendEventCancellationAsync(IEnumerable<string> emails, string link)
+        {
+            var notification = new Notification(
+                GetHtmlPath(MailOptions.EventCancellationHtml),
+                link,
+                "Event Cancellation");
+            return SendEmailAsync(emails, notification);
+        }
+
+        public Task SendEventConfirmationAsync(IEnumerable<string> emails, string link)
+        {
+            var notification = new Notification(
+                GetHtmlPath(MailOptions.EventConfirmationHtml),
+                link,
+                "Event Confirmation");
+            return SendEmailAsync(emails, notification);
+        }
+
+        public Task SendEventInvitationAsync(IEnumerable<string> emails, string link)
+        {
+            var notification = new Notification(GetHtmlPath(MailOptions.EventInvitationHtml), link, "Event Invitation");
+            return SendEmailAsync(emails, notification);
+        }
+
+        public Task SendFriendAcceptAsync(string email, string link)
+        {
+            var notification = new Notification(
+                GetHtmlPath(MailOptions.FriendAcceptHtml),
+                link,
+                "Friend Invitation Accepted");
+            return SendEmailAsync(email, notification);
+        }
+
+        public Task SendFriendInvitationAsync(string email, string link)
+        {
+            var notification = new Notification(
+                GetHtmlPath(MailOptions.FriendInvitationHtml),
+                link,
+                "Friend Invitation");
+            return SendEmailAsync(email, notification);
+        }
+
+        public Task SendPasswordResetAsync(string email, string link)
+        {
+            var notification = new Notification(GetHtmlPath(MailOptions.PasswordResetHtml), link, "Password Reset");
+            return SendEmailAsync(email, notification);
+        }
+
+        private Task SendEmailAsync(IEnumerable<string> emails, Notification notification)
+            => Execute(MailOptions.SendGridApiKey, emails, notification);
+
+        private Task SendEmailAsync(string email, Notification notification)
+        {
+            var emails = new List<string>
+            {
+                email
+            };
+            return SendEmailAsync(emails, notification);
+        }
+
+        private Task Execute(string apiKey, IEnumerable<string> emails, Notification notification)
+        {
+            var client = new SendGridClient(apiKey);
+            var from = new EmailAddress("GameBoard.contact@gmail.com", "GameBoard");
+            var tos = new List<EmailAddress>();
+            var subject = notification.Subject;
+            var htmlContent = notification.Html;
+
+            foreach (var email in emails)
```

Comment:
R# says "LINQ". V0ldek says "LINQ!!!".

Reply:
Done

Reply:
That's not what I or R# had in mind I think. AFAIK using `List.ForEach` is considered bad style to begin with, it's not LINQ, and still you could write
```csharp
var tos = email.Select(e => new EmailAddress(e)).ToList();
```
which is a bit shorter and looks better.

Also, you've left a double endline after line 96, remove one of them.

Reply:
Also also, a thing I haven't considered last time, is the argument `apiKey` needed for this method? You have two methods, `SendEmailAsync` and `SendNotificationAsync`, where the first one does nothing, only passes a value from the class's member field to the second one. I don't see any benefit of doing that, you could get rid of the argument and simply call `SendNotificationAsync` every time (or just remove `SendNotificationAsync` and replace `SendEmailAsync` with its body).

Reply:
Done